elastic4s-http/src/main/scala/com/sksamuel/elastic4s/http/ResponseHandler.scala

```diff
@@ -0,0 +1,67 @@
+package com.sksamuel.elastic4s.http
+
+import java.nio.charset.Charset
+
+import com.fasterxml.jackson.annotation.JsonInclude
+import com.fasterxml.jackson.core.JsonParser
+import com.fasterxml.jackson.databind.{DeserializationFeature, ObjectMapper}
+import com.fasterxml.jackson.module.scala.DefaultScalaModule
+import com.fasterxml.jackson.module.scala.experimental.ScalaObjectMapper
+
+object JacksonSupport {
+
+  val mapper: ObjectMapper with ScalaObjectMapper = new ObjectMapper with ScalaObjectMapper
+  mapper.registerModule(DefaultScalaModule)
+
+  mapper.setSerializationInclusion(JsonInclude.Include.NON_NULL)
+  mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
+  mapper.configure(DeserializationFeature.FAIL_ON_IGNORED_PROPERTIES, false)
+  mapper.configure(DeserializationFeature.ACCEPT_SINGLE_VALUE_AS_ARRAY, true)
+  mapper.configure(JsonParser.Feature.ALLOW_UNQUOTED_FIELD_NAMES, true)
+  mapper.configure(JsonParser.Feature.ALLOW_SINGLE_QUOTES, true)
+}
+
+import org.apache.http.HttpEntity
+import org.elasticsearch.client.{Response, ResponseException}
+
+import scala.io.{Codec, Source}
+import scala.util.{Failure, Try}
+
+trait ResponseHandler[U] {
+  def onResponse(response: Response): Try[U]
+  def onError(e: Exception): Try[U]
+}
+
+// a ResponseHandler that marshalls the body into the required type using Jackson
+// the response body is converted into a string using a codec derived from the content encoding header
+// if the content encoding header is null, then UTF-8 is assumed
+object ResponseHandler {
```

Comment:
The idea behind `ResponseHandler` is that it handles the complete response, not just exceptions. It means that requests that need to handle 404s or other things can be supported by just writing another handler. So if we have a request type that returns text, or a request type that handles 204s specially, or whatever, it can be done easily. 

So its like the original error handler but expanded. 

Of course most of the elastic responses are just "marshall to an object" and so that's what the `DefaultResponseHandler` does.